File: VDText/VDText/VDTextView.m
//
//  VDTextView.m
//  VDText
//
//  Created by Harwyn T'an on 2019/12/12.
//  Copyright © 2020 vvard3n. All rights reserved.
//

#import "VDTextView.h"
#import "VDTextKVO.h"

@interface VDTextView ()// <UITextViewDelegate>

@property (nonatomic, assign) NSRange preSelecteRanget;
@property (nonatomic, assign) BOOL delConform;
@property (nonatomic, strong) VDTextKVO *kvo;

@end

@implementation VDTextView

- (void)resetDelConform {
    self.delConform = NO;
}

-(void)dealloc {
//    [self removeObserver:self forKeyPath:@"selectedRange"];
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (instancetype)initWithFrame:(CGRect)frame {
    if (self = [super initWithFrame:frame]) {
        self.kvo = [VDTextKVO new];
        [self addObserver:self.kvo forKeyPath:@"selectedTextRange" options:NSKeyValueObservingOptionNew context:nil];
        [[NSNotificationCenter defaultCenter]addObserver:self selector:@selector(textDidChange:) name:UITextViewTextDidChangeNotification object:self];
    }
    return self;
}

- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context
{
    if (context == @"selectedRange") {
        
    } else {
        [super observeValueForKeyPath:keyPath ofObject:object change:change context:context];
    }
}

- (void)textDidChange:(NSNotification *)noti {
    NSLog(@"修改文字");
}

//-(void)setPlaceholder:(NSString *)placeholder {
//    _placeholder = placeholder;
//    // setNeedsDisplay会在下一个消息循环时刻，调用drawRect:
//    [self setNeedsDisplay];
//}

-(void)setText:(NSString *)text {
    [super setText:text];
    // setNeedsDisplay会在下一个消息循环时刻，调用drawRect:
    [self setNeedsDisplay];
}

-(void)setFont:(UIFont *)font {
    [super setFont:font];
    // setNeedsDisplay会在下一个消息循环时刻，调用drawRect:
    [self setNeedsDisplay];
}

//- (void)drawRect:(CGRect)rect {
//
//    // 如果有输入文字，就直接返回，不画占位文字
//    if (self.hasText) return;
//    //设置文字属性
//    NSMutableDictionary *attributes = [NSMutableDictionary dictionary];
//    attributes[NSFontAttributeName] = self.font;
//    attributes[NSForegroundColorAttributeName] = self.placeholderColor ? self.placeholderColor : [UIColor lightGrayColor];
//    //画文字
//    CGFloat x = 5;
//    CGFloat width = rect.size.width -2 * x;
//    CGFloat y = 8;
//    CGFloat height = rect.size.height - 2 * y;
//    CGRect placeholderRect = CGRectMake(x, y, width, height);
//    [self.placeholder drawInRect:placeholderRect withAttributes:attributes];
//}


- (BOOL)becomeFirstResponder {
    self.delConform = NO;
    
    return [super becomeFirstResponder];
}

- (BOOL)resignFirstResponder {
    self.delConform = NO;
    
    return [super resignFirstResponder];
}

- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    self.delConform = NO;
    
    [super touchesBegan:touches withEvent:event];
}

#pragma mark - Public Set Get

- (void)deleteBackward {
    NSRange effectiveRange;
    if (self.selectedRange.location == 0) {
        [super deleteBackward];
        return;
    }
    VDTextBinding *binding = [self.attributedText attribute:VDTextBindingAttributeName atIndex:self.selectedRange.location - 1 longestEffectiveRange:&effectiveRange inRange:NSMakeRange(0, self.attributedText.length)];
    if (!binding) {
        [super deleteBackward];
    }
    else {
        NSMutableAttributedString *attrbuteString = self.attributedText.mutableCopy;
        self.preSelecteRanget = self.selectedRange;
        
        if (self.selectedRange.length > 0) {
            NSMutableAttributedString *attrbuteString = self.attributedText.mutableCopy;
            [attrbuteString replaceCharactersInRange:self.selectedRange withString:@""];
            self.attributedText = attrbuteString;
            self.preSelecteRanget = NSMakeRange(self.preSelecteRanget.location, 0);
            self.selectedRange = self.preSelecteRanget;
            if ([self.delegate respondsToSelector:@selector(textViewDidChange:)]) {
                [self.delegate textViewDidChange:self];
            }
            return;
        }
        
        if (!self.delConform) {
            self.delConform = YES;
            
//            [attrbuteString addAttributes:@{NSBackgroundColorAttributeName : [UIColor blueColor]} range:effectiveRange];
            self.preSelecteRanget = effectiveRange;
        }
        else {
            self.delConform = NO;
            [attrbuteString replaceCharactersInRange:effectiveRange withString:@""];
            
            self.preSelecteRanget = effectiveRange;
            self.preSelecteRanget = NSMakeRange(self.preSelecteRanget.location, 0);
        }
        
        self.attributedText = attrbuteString;
        self.selectedRange = self.preSelecteRanget;
    }
}

#pragma mark - UITextViewDelegate
- (BOOL)textViewShouldBeginEditing:(UITextView *)textView {
    if ([self.delegate respondsToSelector:@selector(textViewShouldBeginEditing:)]) {
        return [self.delegate textViewShouldBeginEditing:self];
    }
    return YES;
}

@end

File: VDText/VDText/VDTextView.swift
//
//  VDTextView.swift
//  VDText
//
//  Created by Harwyn T'an on 2019/12/12.
//  Copyright © 2020 vvard3n. All rights reserved.
//

import UIKit

/// 自定义 `UITextView`，主要扩展了 `deleteBackward` 行为：
/// 当光标前一个字符带有 `VDTextBinding` 属性时，删除操作会按"先确认、再整体删除"的方式处理整段 binding 文本。
@objcMembers
public final class VDTextView: UITextView {

    // MARK: - Public

    /// 重置"删除确认"标记。
    ///
    /// `VDTextEditor` 在监听到光标移动时会调用，避免上一次的"待确认删除"状态影响下一次操作。
    public func resetDelConform() {
        delConform = false
    }

    // MARK: - Private state

    private var preSelectedRange: NSRange = NSRange(location: 0, length: 0)
    private var delConform: Bool = false

    // MARK: - Init

    public override init(frame: CGRect, textContainer: NSTextContainer?) {
        super.init(frame: frame, textContainer: textContainer)
    }

    public required init?(coder: NSCoder) {
        super.init(coder: coder)
    }

    // MARK: - Overrides

    public override var text: String! {
        didSet {
            // 与原 OC 行为保持一致：text 变更后请求一次重绘
            setNeedsDisplay()
        }
    }

    public override var font: UIFont? {
        didSet {
            setNeedsDisplay()
        }
    }

    public override func becomeFirstResponder() -> Bool {
        delConform = false
        return super.becomeFirstResponder()
    }

    public override func resignFirstResponder() -> Bool {
        delConform = false
        return super.resignFirstResponder()
    }

    public override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        delConform = false
        super.touchesBegan(touches, with: event)
    }

    /// 处理删除：当光标前一个字符位于一段 `VDTextBinding` 内时，
    /// 第一次按删除会"选中"该 binding 段并把 `delConform` 置 true，
    /// 第二次按删除才真正移除整段 binding。
    public override func deleteBackward() {
        let attributed = attributedText ?? NSAttributedString()
        let selected = selectedRange

        if selected.location == 0 {
            super.deleteBackward()
            return
        }

        var effectiveRange = NSRange(location: 0, length: 0)
        let bindingValue = attributed.attribute(.vdTextBinding,
                                                at: selected.location - 1,
                                                longestEffectiveRange: &effectiveRange,
                                                in: NSRange(location: 0, length: attributed.length))

        guard let _ = bindingValue as? VDTextBinding else {
            delConform = false
            super.deleteBackward()
            return
        }

        let mutable = NSMutableAttributedString(attributedString: attributed)
        preSelectedRange = selected

        // 当前有选区：直接删除选区，恢复 caret
        if selected.length > 0 {
            delConform = false
            mutable.replaceCharacters(in: selected, with: "")
            attributedText = mutable
            preSelectedRange = NSRange(location: preSelectedRange.location, length: 0)
            selectedRange = preSelectedRange
            delegate?.textViewDidChange?(self)
            return
        }

        // 当前是 caret，且位于 binding 内：第一次只把整段 binding "选中"，第二次才真正删除
        if !delConform {
            delConform = true
            preSelectedRange = effectiveRange
        } else {
            delConform = false
            mutable.replaceCharacters(in: effectiveRange, with: "")
            preSelectedRange = NSRange(location: effectiveRange.location, length: 0)
            attributedText = mutable
        }

        selectedRange = preSelectedRange
    }
}
